### DISCUSSION TITLE
How to generate struct methods to frontend in this case?

### DISCUSSION THREAD

**Discussion by KiddoV** [wailsapp/wails#3314] [Q&A]
Why is my generated JS class doesn't come with my struct method in this case?

I have a every basic struct:
```go
package test

type TestStruct struct {
    Field1 string
    Field2 int
}

func (t *TestStruct) SomeMethod() error {
    fmt.Println("TestStruct.SomeMethod() called!")
    return nil
}
```

I have a function to init this struct in the main ``app`` struct:
```go
package main

func (a *App) NewTestStruc() *test.TestStruct {
    return &test.TestStruct{}
}
```
------
Now in the frontend I call the function to init the struct:
```ts
import { NewTestStruc } from 'wailsjs/go/main/App';
import { test } from 'wailsjs/go/models';

[...]
var testStructFE: test.TestStruct;

NewTestStruc().then(st => {
    testStructFE = st;
    //Now my `testStructFE` var doesn't come with any methods
}).catch(() =>{});

```

However I don't see any struct methods created in the `models.ts` with class `test` in my frontend.

Is it even possible for Wails?

**Comment by KiddoV** [comment]
Answer in https://discord.com/channels/1042734330029547630/1061396491769479300/1217547615537729688
I don't think it is possible.

**Reply by leaanthony** [reply]
Is it because you don't have json tags?

**Reply by KiddoV** [reply]
I do have all of my JSON tags in the struct fields; the code provided was just a minimal example of my actual code. However, I've never placed my struct instance directly in the Bind. Instead, I intended to instantiate it within an `app method`, store it in a JavaScript variable, and then call its methods from the frontend. I've come to realize that in Wails, I only have two options:
1. Instantiate a struct instance first in Go side and include it in the Bind to expose all its methods in the frontend (though its struct fields wont be exposed), or
2. Bind the struct to the frontend via an `app` method to access the struct fields (as a class type in JavaScript, but without its methods).

There is noway to get both (the struct fields and methods as a class in JS)

**Comment by geocine** [comment]
@KiddoV  you can use `Bind` on `wails.Run` .  You may check https://wails.io/docs/guides/application-development/#dealing-with-context-when-binding-multiple-structs 

Your struct

```golang

package test

type TestStruct struct {
    Field1 string
    Field2 int
}

func (t *TestStruct) SomeMethod() error {
    fmt.Println("TestStruct.SomeMethod() called!")
    return nil
}

func NewTestStruc() *TestStruct {
    return &TestStruct{}
}
```

Then on your `app.go`

```golang
package main

import (
	"yourpkg/test"
)

type App struct {
	ctx context.Context

	TestStruct *TestStruct
}

func NewApp() *App {

	app := &App{
		TestStruct: test.NewTestStruc(),
	}
	
	return app
}
```

Then on you `main.go`

```golang
func main() {
	// ....
	app := NewApp()

	// Create application with options
	err := wails.Run(&options.App{
		// ....
		Bind: []interface{}{
			app,
			app.TestStruct, // Your Struct
		},
	})
}

```
This should generate 

`wails/go/test/TestStruct`

**Reply by KiddoV** [reply]
Thanks for the answer, but as I [said earlier](https://github.com/wailsapp/wails/discussions/3314#discussioncomment-8816610) in ``1``. If you do this, you can only get the struct methods generated, without the struct fields exposed in the frontend.

I would expected to see something like this in frontend:
``wails/go/test/..`` or ``wails/go/models``
```ts
export class TestStruct {
    //The struct fields...
    field1: string;
    field2: number;
    
    [...]
    //...Then the method:
    SomeMethod(): Promise<void>;
}
```
However, I was mistaken. This isn't possible in ``Wails``, or at least not how ``Wails binding generation`` works.
You can only instantiate ``a struct instance`` in the backend then exposed its methods to frontend via ``&options.App.Bind`` (without its struct fields as an exported class). Or you can use the struct as ``a type`` for some method in backend, and it will automatically generate ``exported class`` in the frontend (without its methods).

**Reply by geocine** [reply]
Right I just noticed, it doesn't export its as a class but a module.
